### ISSUE TITLE
Add rule to enforce new line for every parameter in a multi-line function call

### ISSUE THREAD

**Issue by cheeZery** [eslint/eslint#10323]
**Please describe what the rule should do:**

The rule should enforce one line per parameter for multi-line function calls.

**What category of rule is this? (place an "X" next to just one item)**

[X] Enforces code style
[X] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// Bad
someFunction(
    true, "abc",
    123
);

// Good
someFunction(
    true, 
    "abc",
    123
);

// Good
someFunction(
    true
);

// All parameters in one line should also be fine (imho)
someFunction(
    true, "abc", 123
);
```

**Why should this rule be included in ESLint (instead of a plugin)?**
In multi-line function calls with more than one parameter it can be confusing if some parameters share a line and others have an own line. Especially if some parameters are very long and some function arguments are optional.

```js
function doSomething(someBool, someText = "foo", someOtherText = "bar") {
    if (someBool) {
        console.log(someText);
    } else {
        console.log(someOtherText);
    }
}

// At first glance one might think, "oh geez" is used for someText
doSomething(
    $(".someElement").hasClass("barz") && !$(".someElement").hasClass("fooz"), "blah",
    "oh geez" 
);
```

**Comment by platinumazure** [comment]
Hi @cheeZery, thanks for the issue!

I think it might be good to model this after `object-property-newline` and/or `array-element-newline` and have some or all of the same options. (For example, a "consistent" option could be used for all or nothing, and we could have minimum/maximum parameter counts to force newline requirements, etc.)

I'll champion this.

**Comment by platinumazure** [comment]
@eslint/eslint-team Could we get one more :+1: on this? (If not, it might be time to close.)

**Comment by Aladdin-ADD** [comment]
now it's accepted. 😄

**Comment by kaicataldo** [comment]
It looks like this was accepted, but the implementation details of the proposal still need to be worked out. How do we want to proceed here? cc @platinumazure

**Comment by platinumazure** [comment]
Hi @kaicataldo, I had envisioned that this rule would have a similar set of options as `object-property-newline` and `array-element-newline` since this rule is basically analogous to those rules for the function parameter case.

I do think we should be careful to distinguish between the function definition and function call cases, though. Possibly we should have separate rules for those. This proposal seems to be about the function call case.

**Comment by finico** [comment]
@platinumazure May I get this issue to implement?

I have some questions:

What's the name of the new rule might be?
`func-call-newline` or `func-call-args-newline` or `args-newline` or something other?

Should it check `NewExpression`?

Which is better?
1. It has `"always" | "never" | "consistent"` options.
2. It checks like "consistent" without any options and `max-len` rule will enforce to add new lines.

Should it check spaces and new lines between parentheses and arguments?
```js
fn(
    "first", "second");
```
Is there a rule for that? (I couldn't find)

There are some cases that Prettier does, e.g.
```js
fn(1, true, {
    first: "value",
    second: "value"
});

fn(1, true, [
    "first", 
    "second"
]);
```
Should it be valid? Maybe the rule needs an option for this?
Or we can always check only the first token's line, then it will be valid.

**Comment by platinumazure** [comment]
Hi @finico, thanks for volunteering and sorry for missing your comment up until now.

In keeping with `array-bracket-newline` and `array-element-newline`, I would suggest we follow a similar general design:

* Let's separate the checks for newlines between function parentheses and arguments (akin to `array-bracket-newline`) and checks for newlines between arguments (akin to `array-element-newline`) in two different rules
* The name of this rule could be something like `function-call-argument-newline`
* I think it makes sense to include `NewExpression`. We could add options later for configuring different behavior between `NewExpression` and `CallExpression`.
* I think starting with similar options as `array-element-newline` makes a lot of sense. Don't feel that you need to implement all of the options in one PR though. I would suggest implementing the basic and/or most likely to be used options. If the team feels that some option is critical for the first release of the rule, we could work with you during PR review.

These are just suggestions, let me know what you think. Also, if any team members have different thoughts on how to proceed, let's continue the discussion. Thanks again!

**Comment by finico** [comment]
Hello @platinumazure, it took a while, but I did it.
I didn't immediately realize what was wrong with the commit message, so there's a little bit of a mess.

**Comment by kaicataldo** [comment]
It looks like this has been completed!
